### PR TITLE
Framework for server config validator

### PR DESCRIPTION
**Pull request by jbu**
Framework for the server validator work. see #417.

### CODE REVIEW

File: ga4gh/cli.py [hunk 1]
```diff
@@ -816,3 +820,72 @@ def client_main(parser=None):
...
+    else:
+        return None
+
+
+def configtest_main(parser=None):
+    test_config = TestConfigExists(parser)
+    test_config.setUp()
+    tests = ['test_config_exists']
+    results = map(lambda t: run_test(test_config, t), tests)
+    print('{0}/{1} tests passed'.format(results.count(None), len(results)))
```
Comment (dcolligan) [inline ga4gh/cli.py]: we shouldn't have any `print` calls in the ga4gh module code ... these should be calls to some kind of logging tool (even if it is something as simple as calling a method that just calls print)

File: ga4gh/cli.py [hunk 2]
```diff
@@ -816,3 +820,72 @@ def client_main(parser=None):
...
+
+    def test_config_exists(self):
+        self.assertIsInstance(self.app.config["REQUEST_VALIDATION"],
+                              types.BooleanType,
+                              'REQUEST_VALIDATION is not a boolean')
+        self.assertIsNotNone(self.app.config["DATA_SOURCE"],
+                             'DATA_SOURCE must be set')
+
+
+def run_test(test_obj, test):
```
Comment (dcolligan) [inline ga4gh/cli.py]: we use camelCase variable naming conventions in our code rather than this_style

File: ga4gh/cli.py [hunk 3]
```diff
@@ -816,3 +820,72 @@ def client_main(parser=None):
...
+    line arguments) we need to break the usual unittest pattern and instead
+    implement our own simple test runner.
+    """
+
+    def __init__(self, parser=None):
+        self.parser = parser
+
+    def setUp(self):
+        self.app = flask.Flask(__name__)
+        parser = self.parser if self.parser is not None \
```
Comment (dcolligan) [inline ga4gh/cli.py]: @jeromekelleher claims that line-continuation backslashes kill puppies, so I would avoid this

I don't know if we have an opinion on ternary operators.  Personally, I avoid them, but I'd be willing to hear what other people think...

File: ga4gh/cli.py [hunk 4]
```diff
@@ -816,3 +820,72 @@ def client_main(parser=None):
...
+    However, as some of the tests require state injection (through command
+    line arguments) we need to break the usual unittest pattern and instead
+    implement our own simple test runner.
+    """
+
+    def __init__(self, parser=None):
+        self.parser = parser
+
+    def setUp(self):
+        self.app = flask.Flask(__name__)
```
Comment (dcolligan) [inline ga4gh/cli.py]: I don't know if the configuration checker should ever start up a Flask instance; that seems outside the scope of what it's supposed to do...

File: ga4gh/cli.py [hunk 5]
```diff
@@ -816,3 +820,72 @@ def client_main(parser=None):
...
+            else argparse.ArgumentParser(
+                description="GA4GH server configuration validator")
+        parser.add_argument(
+            "--config", "-c", default='DevelopmentConfig', type=str,
+            help="The configuration to use")
+        parser.add_argument(
+            "--config-file", "-f", type=str, default=None,
+            help="The configuration file to use")
+        args = parser.parse_args()
+        config_str = 'ga4gh.serverconfig:{0}'.format(args.config)
```
Comment (dcolligan) [inline ga4gh/cli.py]: should be a camelCase var ... same throughout the rest of your changes

File: ga4gh/cli.py (hunk 3 above)
Comment (jbu) [inline ga4gh/cli.py]: It kills puppies, but it passes the pep8 linter.

File: ga4gh/cli.py (hunk 4 above)
Comment (jbu) [inline ga4gh/cli.py]: The idea is to reuse the config reader in the flask app. I could do that more manually, but might end up either using flask internals, or re-writing code. Another alternative is to pull the config reader out of the app into something more standalone and test it there.

File: ga4gh/cli.py (hunk 5 above)
Comment (jbu) [inline ga4gh/cli.py]: righto.

File: ga4gh/cli.py (hunk 1 above)
Comment (jbu) [inline ga4gh/cli.py]: adding the logging module. Will print to stdout to start with.

File: ga4gh/cli.py (hunk 2 above)
Comment (jbu) [inline ga4gh/cli.py]: So no 'N806 variable in function should be lowercase' from flask8?

File: ga4gh/cli.py (hunk 3 above)
Comment (jbu) [inline ga4gh/cli.py]: Happy to be guided on ternary operators. I found them more readable once I got used to them, particularly as it becomes obvious where and when a variable is being assigned.

File: ga4gh/cli.py (hunk 2 above)
Comment (jeromekelleher) [inline ga4gh/cli.py]: We adopted camelCase for this project because that's the convention that the GA4GH protocol uses. I would have preferred snake_case, but having mixed conventions is much worse than one you don't like. I'm not sure where flake8 stands at the moment with camelCase identifiers, if it decides to start checking this, we'll need to update our Travis script.

File: ga4gh/cli.py (hunk 3 above)
Comment (jeromekelleher) [inline ga4gh/cli.py]: The pep8 tool doesn't _currently_ have any issues with backslash line continuations, but pep8 is actually pretty clear on this:

> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line  continuation.

So, I think we should avoid backslashes where we can. I think this is such an occasion. What about
```python
    parser = (
        self.parser 
        if self.parser is not None 
        else argparse.ArgumentParser(
            description="GA4GH server configuration validator"))
```
I quite like ternary operators, but I find it hard to write PEP8 compliant code using them, for just this reason. I would tend to favour the simpler
```python
    parser = self.parser
    if parser is None:
          parser = ArgumentParser(...)
```
in this situation.

**Comment by jeromekelleher** [comment]
Looks good, thanks @jbu. I agree we should reuse the Flask configuration parsers to get our data in, otherwise we'll be duplicating a lot of their code. Reusing the unittest framework to do the tests is very sneaky - I like it! 

The only issue I have here is that I think we should take the real work out of the 'cli' module. Can you move all the code where we do the actual testing and so on into a new module called 'ga4gh/configtest.py' please?

**Comment by jbu** [comment]
So I found there was a module configuration test that wasn't passing. I added configtest.py to the notional 'libraries' module as it's the best fit currently. Or, I could start a new module.

File: ga4gh/configtest.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
+from __future__ import division
```
Comment (jeromekelleher) [inline ga4gh/configtest.py]: Can we get a module docstring here please? Doesn't have to be anything elaborate, just a quick description of what the module is for.

File: ga4gh/configtest.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import argparse
```
Comment (jeromekelleher) [inline ga4gh/configtest.py]: I'd be happier if we could keep all of the argparse stuff in the cli module. The reasoning for this is that it makes it a lot easier for us to keep our CLI applications consistent if we try to share as much code as possible between then. There's a quite a lot of sharing between the apps at the moment, as we use functions to create command line parsers that have the same arguments and so on.

File: ga4gh/configtest.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+class TestConfigExists(unittest.TestCase):
+    """This class implements configuration checks for the GA4GH server. It is
+    structured as a set of unit tests, and relies on the unittest module.
+    However, as some of the tests require state injection (through command
+    line arguments) we need to break the usual unittest pattern and instead
+    implement our own simple test runner.
+    """
+
+    def __init__(self, parser=None):
```
Comment (jeromekelleher) [inline ga4gh/configtest.py]: Following on from the point above, how about we pass configFile and configClass as arguments to this constructor, and take the command line argument parsing bit out of the setUp function below?

File: ga4gh/cli.py [hunk 9]
```diff
@@ -816,3 +818,37 @@ def client_main(parser=None):
...
+    try:
+        getattr(testObj, test)()
+    except AssertionError as e:
+        return test, e
+    else:
+        return None
+
+
+def configtest_main(parser=None):
+    testConfig = configtest.TestConfig(parser)
```
Comment (jeromekelleher) [inline ga4gh/cli.py]: Maybe we could use `unittest.loadTestsFromModule` or similar to discover the tests, so we don't have to register each one?

Comment (jbu) [inline ga4gh/cli.py]: I rejected that originally, but perhaps prematurely. Will have another look.

File: ga4gh/cli.py [hunk 10]
```diff
@@ -816,3 +818,59 @@ def client_main(parser=None):
...
+    except Exception as e:
+        results = [('setUp failed', e)]
+    else:
+        results = run_tests(testConfig, tests)
+
+    logging.basicConfig(level=logging.INFO)
+    log = logging.getLogger(__name__)
+    log.info('{0}/{1} tests passed'.format(results.count(None),
+                                           len(results)))
+    for r in results:
```
Comment (dcolligan) [inline ga4gh/cli.py]: we tend to frown on single-letter variables here, unless they are counting variables (which this one is not)

File: ga4gh/cli.py [hunk 11]
```diff
@@ -816,3 +818,59 @@ def client_main(parser=None):
...
+    configStr = 'ga4gh.serverconfig:{0}'.format(args.config)
+    testConfig = configtest.TestConfig(configStr,
+                                       'GA4GH_CONFIGURATION',
+                                       args.config_file)
+    tests = ['test_config_parameters']
+    try:
+        # setUp can fail, so we wrap it in a try to help it look like
+        # the other tests.
+        testConfig.setUp()
+    except Exception as e:
```
Comment (dcolligan) [inline ga4gh/cli.py]: another single-letter variable

File: ga4gh/cli.py [hunk 12]
```diff
@@ -816,3 +818,59 @@ def client_main(parser=None):
...
+
+def run_tests(testObj, tests):
+    """
+    Run a list of tests on a testObj
+    :param testObj: The object under test
+    :param tests: the list of test methods to run
+    :return: a list of results, either None, or a tuple of the test name and
+    reason for failure (raised exception)
+    """
+    def run_test(o, t):
```
Comment (dcolligan) [inline ga4gh/cli.py]: single-letters...

File: ga4gh/cli.py [hunk 13]
```diff
@@ -816,3 +818,59 @@ def client_main(parser=None):
     else:
         runner = args.runner(args)
         runner.run()
+
+##############################################################################
+# Configuration testing
+##############################################################################
+
+
+def run_tests(testObj, tests):
```
Comment (jeromekelleher) [inline ga4gh/cli.py]: All functions should be camelCase also (the '_main' functions were the only exception to this, which I keep forgetting to fix!). Sorry, should have spotted this earlier.

File: ga4gh/configtest.py (hunk 6 above)
Comment (jbu) [inline ga4gh/configtest.py]: done. I don't know why it's not showing here.

**Comment by jbu** [comment]
have added test discovery. I had a good look at how one might pass state to unit tests using the unittest module, and the best I could come up with was to just set class variables in the test class before running any tests. There were potentially other ways involving subclassing bits of the test framework, but they started to get into the framework internals. None of the options are fantastic, but at least this one is simple.

**Comment by jeromekelleher** [comment]
I like it @jbu, very tidy. Shall we squash these commits and aim to merge this much, or would you prefer to add some more functionalilty first?

**Comment by jbu** [comment]
I'll squash these (once travis goes green). We can then build on that.

**Comment by jbu** [comment]
Ready for merge.

**Comment by jeromekelleher** [comment]
+1.

@dcolligan, can you merge please, if you are happy with this?

File: ga4gh/cli.py [hunk 14]
```diff
@@ -816,3 +821,62 @@ def client_main(parser=None):
...
+##############################################################################
+# Configuration testing
+##############################################################################
+
+
+class SimplerResult(unittest.TestResult):
+    """
+    The TestResult class gives formatted tracebacks as error messages, which
+    is not what we want. Instead we just want the error message from the
+    err praram. Hence this subclass.
```
Comment (pgrosu) [inline ga4gh/cli.py]: Probably might be helpful to substitute `praram` with `param`.
